### PR TITLE
blueman: fix python wrapper mockup

### LINKED ISSUE #28260: Blueman applet does not start (incomplete python path)
## Issue description

When starting blueman applet, the command fails with `ImportError: No module named blueman.Constants` (see below). The issue can be fixed by exporting first `PYTHONPATH=${blueman}/lib/python2.7/site-packages` before invoking blueman-applet

The applet file is so much wrapped that it is nearly impossible to understand what happens here.

```
Here are all the redrections to reach blueman-applets
/nix/store/...-blueman-2.0.4/bin/blueman-applet
/nix/store/...-blueman-2.0.4/bin/.blueman-applet-wrapped_
/nix/store/...-blueman-2.0.4/bin/.blueman-applet-wrapped
/nix/store/...-blueman-2.0.4/bin/..blueman-applet-wrapped-wrapped
```

### Steps to reproduce

install blueman-applet, and try to run ${blueman-applet}/bin/blueman-applet

## Technical details

* System: 17.09pre111460.b636a98efb (custom branch, based on d4ef5ac0e962bd6604dda38617c4b98c77a62949)
* Nix version: nix-env (Nix) 1.11.13
* Nixpkgs version: 17.09pre111460.b636a98efb (as above)
* Sandboxing enabled: yes

The complete error :
```
$ blueman-applet 
Traceback (most recent call last):
  File "/nix/store/njz24q9j9xkhs3m97i8ja6zipzq6214x-blueman-2.0.4/bin/..blueman-applet-wrapped-wrapped", line 15, in <module>
    from blueman.Constants import *
ImportError: No module named blueman.Constants
```

/cc @abbradar as maintainer of blueman.

### PR DESCRIPTION
**Pull request by layus**
The wrapper missed $out in the python path, producing weird errors where
blueman could not find itself :-).

Fixes #28260.
/cc @abbradar as potential reviewer.

### CODE REVIEW

File: pkgs/tools/bluetooth/blueman/default.nix [hunk 1]
```diff
@@ -31,10 +31,11 @@ in stdenv.mkDerivation rec {
 
   configureFlags = [ (lib.enableFeature withPulseAudio "pulseaudio") ];
 
-  preFixup = ''
+  # This mimics ../../../development/interpreters/python/wrap.sh
+  postFixup = ''
     makeWrapperArgs="--prefix PATH ':' ${binPath}"
```
Comment (layus) [inline pkgs/tools/bluetooth/blueman/default.nix]: Well, it is because blueman calls xdg-utils and dhcp with `system()` invocation.
I admit that it is hacky, but wrapping a third time the same executable looks even worse.

Comment (FRidh) [inline pkgs/tools/bluetooth/blueman/default.nix]: Yes, but I thought wrapPythonPrograms adds `$out` to `PATH` already.

Comment (abbradar) [inline pkgs/tools/bluetooth/blueman/default.nix]: @FRidh `wrapPythonPrograms` does but `wrapPythonProgramsIn` doesn't. We need the latter because `/libexec` also needs to be wrapped (perhaps we could just add it to `wrapPythonPrograms` instead?).

File: pkgs/tools/bluetooth/blueman/default.nix [hunk 2]
```diff
@@ -31,10 +31,11 @@ in stdenv.mkDerivation rec {
 
   configureFlags = [ (lib.enableFeature withPulseAudio "pulseaudio") ];
 
-  preFixup = ''
+  postFixup = ''
```
Comment (abbradar) [inline pkgs/tools/bluetooth/blueman/default.nix]: Why this change?

File: pkgs/tools/bluetooth/blueman/default.nix (hunk 1 above)
Comment (FRidh) [inline pkgs/tools/bluetooth/blueman/default.nix]: @abbradar ah I forgot about that. Awhile ago I restricted `wrapPythonPrograms` to look only in `$out/bin` instead of the whole `$out`. Maybe it would make sense to revert that or indeed explicitly include `$out/libexec`.

Comment (abbradar) [inline pkgs/tools/bluetooth/blueman/default.nix]: @FRidh Explicitly adding `libexec` sounds good to me (BTW, do you remember why the change was made in the first place?)

Comment (FRidh) [inline pkgs/tools/bluetooth/blueman/default.nix]: https://github.com/NixOS/nixpkgs/commit/c10af9e744c91dff1ccc07a52a0b57d1e4d339f3

My reasoning was that executables belong in `/bin` and not elsewhere. I did not think of `libexec` at the time.

File: pkgs/tools/bluetooth/blueman/default.nix (hunk 2 above)
Comment (layus) [inline pkgs/tools/bluetooth/blueman/default.nix]: Because this package includes wrapGAppsHooks. If you wrap in preFixup, then wrapGAppsHooks wraps both the executable and its wrapper. This gives the insane wrappers chain described in #28260.

Now that I write it, I wonder why the same does not occur with wrapPythonProgramsIn. Maybe because it only wraps python packages, not wrappers ?

File: pkgs/tools/bluetooth/blueman/default.nix (hunk 1 above)
Comment (layus) [inline pkgs/tools/bluetooth/blueman/default.nix]: Just my two cents here, but wrapsPythonProgramsIn is the right place to add $out to the pythonPath. The interface is error prone otherwise. May I propose 
```diff
diff --git a/pkgs/development/interpreters/python/wrap.sh b/pkgs/development/interpreters/python/wrap.sh
index 1c74e612b5..f86738f4a0 100644
--- a/pkgs/development/interpreters/python/wrap.sh
+++ b/pkgs/development/interpreters/python/wrap.sh
@@ -1,7 +1,7 @@
 # Wrapper around wrapPythonProgramsIn, below. The $pythonPath
 # variable is passed in from the buildPythonPackage function.
 wrapPythonPrograms() {
-    wrapPythonProgramsIn "$out/bin" "$out $pythonPath"
+    wrapPythonProgramsIn "$out/bin"
 }
 
 # Builds environment variables like PYTHONPATH and PATH walking through closure
@@ -41,7 +41,7 @@ patchPythonScript() {
 # suffix).
 wrapPythonProgramsIn() {
     local dir="$1"
-    local pythonPath="$2"
+    local pythonPath="${2:-$out $pythonPath}"
     local f
 
     buildPythonPath "$pythonPath"
diff --git a/pkgs/tools/bluetooth/blueman/default.nix b/pkgs/tools/bluetooth/blueman/default.nix
index 8512cd4fcb..60686764ec 100644
--- a/pkgs/tools/bluetooth/blueman/default.nix
+++ b/pkgs/tools/bluetooth/blueman/default.nix
@@ -34,8 +34,8 @@ in stdenv.mkDerivation rec {
   postFixup = ''
     makeWrapperArgs="--prefix PATH ':' ${binPath}"
     # This mimics ../../../development/interpreters/python/wrap.sh
-    wrapPythonProgramsIn "$out/bin" "$out $pythonPath"
-    wrapPythonProgramsIn "$out/libexec" "$out $pythonPath"
+    wrapPythonProgramsIn "$out/bin"
+    wrapPythonProgramsIn "$out/libexec"
   '';
 
   meta = with lib; {
```
or maybe `local pythonPath="$out ${2:-$pythonPath}"` ?

**Comment by layus** [comment]
@FRidh, @abbradar : Thanks for your reviews! I replied to above comments.
I think this is ready to land, the other improvements to python infra can still be implemented & applied later.
